File: Top-Interview-Questions-Leetcode/Easy-Collections/dynamic-programming/maximum-subarray-sum.cpp
#include <iostream>
#include <vector>

class Solution {
private:
    
    int divide_and_conquer(std::vector<int>& nums, int low, int high) {
        // Update ...
    }
    
public:
    int maxSubArray(std::vector<int>& nums) {
        int mn = 0, mx = nums[0], sum = 0;
        
        for (int num : nums) {
            sum += num;
            mx = std::max(mx, sum - mn);
            mn = std::min(mn, sum);
        }
        
        return mx;
    }
};

File: Top-Interview-Questions-Leetcode/Easy-Collections/dynamic-programming/maximum-subarray-sum.py
from typing import List

class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        mn, mx, sum = 0, nums[0], 0
        
        for num in nums:
            sum += num
            mx = max(mx, sum - mn)
            mn = min(mn, sum)
        
        return mx